fail gracefully if S3 input files are in Glacier

`check_input_paths` should check if input files are in Glacier and issue a reasonable error message if so.

Here's the error you get when trying to read a glacier-archived input file:

```
Exception running child : java.io.IOException: com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.services.s3.model.AmazonS3Exception: The operation is not valid for the object's storage class (Service: Amazon S3; Status Code: 403; Error Code: InvalidObjectState; Request ID: 25831B4146C8645E; S3 Extended Request ID: n73unxX/a03XRdy/AfGReQm6/Ad4RgF8RNFFh7QlLyK9Xz/SxeGPcaN01HgehimBmy7PS15f1/0=), S3 Extended Request ID: n73unxX/a03XRdy/AfGReQm6/Ad4RgF8RNFFh7QlLyK9Xz/SxeGPcaN01HgehimBmy7PS15f1/0=
	at com.amazon.ws.emr.hadoop.fs.s3n.Jets3tNativeFileSystemStore.handleAmazonServiceException(Jets3tNativeFileSystemStore.java:448)
	at com.amazon.ws.emr.hadoop.fs.s3n.Jets3tNativeFileSystemStore.retrievePair(Jets3tNativeFileSystemStore.java:476)
	at com.amazon.ws.emr.hadoop.fs.s3n.Jets3tNativeFileSystemStore.retrievePair(Jets3tNativeFileSystemStore.java:453)
	at com.amazon.ws.emr.hadoop.fs.s3n.S3NativeFileSystem.open(S3NativeFileSystem.java:1127)
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:790)
	at com.amazon.ws.emr.hadoop.fs.EmrFileSystem.open(EmrFileSystem.java:192)
	at org.apache.hadoop.mapred.LineRecordReader.<init>(LineRecordReader.java:109)
	at org.apache.hadoop.mapred.TextInputFormat.getRecordReader(TextInputFormat.java:67)
	at org.apache.hadoop.mapred.MapTask$TrackedRecordReader.<init>(MapTask.java:170)
	at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:433)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:344)
	at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:175)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1840)
	at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:169)
Caused by: com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.services.s3.model.AmazonS3Exception: The operation is not valid for the object's storage class (Service: Amazon S3; Status Code: 403; Error Code: InvalidObjectState; Request ID: 25831B4146C8645E; S3 Extended Request ID: n73unxX/a03XRdy/AfGReQm6/Ad4RgF8RNFFh7QlLyK9Xz/SxeGPcaN01HgehimBmy7PS15f1/0=), S3 Extended Request ID: n73unxX/a03XRdy/AfGReQm6/Ad4RgF8RNFFh7QlLyK9Xz/SxeGPcaN01HgehimBmy7PS15f1/0=
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1632)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1304)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1058)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4365)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4312)
	at com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.services.s3.AmazonS3Client.getObject(AmazonS3Client.java:1415)
	at com.amazon.ws.emr.hadoop.fs.s3.lite.call.GetObjectCall.perform(GetObjectCall.java:22)
	at com.amazon.ws.emr.hadoop.fs.s3.lite.call.GetObjectCall.perform(GetObjectCall.java:9)
	at com.amazon.ws.emr.hadoop.fs.s3.lite.executor.GlobalS3Executor.execute(GlobalS3Executor.java:91)
	at com.amazon.ws.emr.hadoop.fs.s3.lite.AmazonS3LiteClient.invoke(AmazonS3LiteClient.java:185)
	at com.amazon.ws.emr.hadoop.fs.s3.lite.AmazonS3LiteClient.getObject(AmazonS3LiteClient.java:102)
	at com.amazon.ws.emr.hadoop.fs.s3n.Jets3tNativeFileSystemStore.retrievePair(Jets3tNativeFileSystemStore.java:466)
	... 14 more
```

--- a/mrjob/emr.py
+++ b/mrjob/emr.py
@@ -667,7 +667,7 @@ class EMRJobRunner(HadoopInTheCloudJobRunner, LogInterpretationMixin):
         local filesystem.
         """
         if self._fs is None:
-            s3_fs = S3Filesystem(
+            self._s3_fs = S3Filesystem(
                 aws_access_key_id=self._opts['aws_access_key_id'],
                 aws_secret_access_key=self._opts['aws_secret_access_key'],
                 aws_session_token=self._opts['aws_session_token'],
@@ -684,11 +684,13 @@ class EMRJobRunner(HadoopInTheCloudJobRunner, LogInterpretationMixin):
 
                 # put _hadoop_fs last because it tries to handle all URIs
                 self._fs = CompositeFilesystem(
-                    self._ssh_fs, s3_fs, self._hadoop_fs, LocalFilesystem())
+                    self._ssh_fs, self._s3_fs, self._hadoop_fs,
+                    LocalFilesystem())
             else:
                 self._ssh_fs = None
                 self._hadoop_fs = None
-                self._fs = CompositeFilesystem(s3_fs, LocalFilesystem())
+                self._fs = CompositeFilesystem(
+                    self._s3_fs, LocalFilesystem())
 
         return self._fs
 
@@ -733,6 +735,28 @@ class EMRJobRunner(HadoopInTheCloudJobRunner, LogInterpretationMixin):
             self._hadoop_fs.set_hadoop_bin([])
 
         self._launch_emr_job()
+
+    def _check_input_path(self, path):
+        """Add a custom check for S3 paths to ensure they're not in
+        Glacier (which causes a cryptic error). See #1887."""
+        # handle non-S3 paths the usual way
+        if not is_s3_uri(path):
+            super(EMRJobRunner, self)._check_input_path(path)
+            return
+
+        exists = False
+
+        for uri, obj in self._s3_fs._ls(path):
+            exists = True
+
+            if obj.storage_class == 'GLACIER':
+                raise IOError(
+                    '%s is archived in Glacier and'
+                    ' cannot be read as input!' % uri)
+
+        if not exists:
+            raise IOError(
+                'Input path %s does not exist!' % (path,))
 
     def _check_output_not_exists(self):
         """Verify the output path does not already exist. This avoids

--- a/mrjob/runner.py
+++ b/mrjob/runner.py
@@ -1016,15 +1016,21 @@ class MRJobRunner(object):
             return
 
         for path in self._input_paths:
-            if path == '-':
-                    continue  # STDIN always exists
+            self._check_input_path(path)
 
-            if not self.fs.can_handle_path(path):
-                continue  # e.g. non-S3 URIs on EMR
+    def _check_input_path(self, path):
+        """Raise :py:class:`IOError` if the given input does not exist or
+        is otherwise invalid. Override this to provide custom check
+        behavior."""
+        if path == '-':
+            return  # STDIN always exists
 
-            if not self.fs.exists(path):
-                raise IOError(
-                    'Input path %s does not exist!' % (path,))
+        if not self.fs.can_handle_path(path):
+            return  # no way to check (e.g. non-S3 URIs on EMR)
+
+        if not self.fs.exists(path):
+            raise IOError(
+                'Input path %s does not exist!' % (path,))
 
     def _add_input_files_for_upload(self):
         """If there is an upload manager, add input files to it."""

--- a/tests/mock_boto3/case.py
+++ b/tests/mock_boto3/case.py
@@ -85,10 +85,11 @@ class MockBoto3TestCase(SandboxedTestCase):
 
         self.start(patch.object(time, 'sleep'))
 
-    def add_mock_s3_data(self, data, age=None, location=None):
+    def add_mock_s3_data(self, data,
+                         age=None, location=None, storage_class=None):
         """Update self.mock_s3_fs with a map from bucket name
         to key name to data."""
-        add_mock_s3_data(self.mock_s3_fs, data, age, location)
+        add_mock_s3_data(self.mock_s3_fs, data, age, location, storage_class)
 
     def add_mock_ec2_image(self, image):
         """Add information about a mock EC2 Image (AMI) to be returned by

--- a/tests/mock_boto3/s3.py
+++ b/tests/mock_boto3/s3.py
@@ -30,12 +30,12 @@ class MockS3Client(object):
     """Mock out boto3 S3 client
 
     :param mock_s3_fs: Maps bucket name to a dictionary with the keys *keys*
-                       and *location*. *keys* maps key name to tuples of
-                       ``(data, time_modified)``. *data* is bytes, and
-                        *time_modified* is a UTC
-                        :py:class:`~datetime.datetime`. *location* is an
-                        optional location constraint for the bucket
-                        (a region name).
+                       and *location*. *keys* maps key name to dicts with
+                       the keys *body* and *time_modified*. *body* is bytes,
+                       and *time_modified* is a UTC
+                       :py:class:`~datetime.datetime`. *location* is an
+                       optional location constraint for the bucket
+                       (a region name).
     """
     def __init__(self,
                  mock_s3_fs,
@@ -107,7 +107,8 @@ def add_mock_s3_data(mock_s3_fs, data, age=None, location=None):
         for key_name, key_data in key_name_to_bytes.items():
             if not isinstance(key_data, bytes):
                 raise TypeError('mock s3 data must be bytes')
-            bucket['keys'][key_name] = (key_data, time_modified)
+            bucket['keys'][key_name] = dict(
+                body=key_data, time_modified=time_modified)
 
         if location is not None:
             bucket['location'] = location
@@ -260,7 +261,8 @@ class MockS3Object(object):
         if not isinstance(data, bytes):
             raise TypeError('Body or Body.read() must be bytes')
 
-        mock_keys[self.key] = (data, _boto3_now())
+        mock_keys[self.key] = dict(
+            body=data, time_modified=_boto3_now())
 
     def upload_file(self, path, Config=None):
         if self.bucket_name not in self.meta.client.mock_s3_fs:
@@ -272,7 +274,8 @@ class MockS3Object(object):
 
         mock_keys = self._mock_bucket_keys('PutObject')
         with open(path, 'rb') as f:
-            mock_keys[self.key] = (f.read(), _boto3_now())
+            mock_keys[self.key] = dict(
+                body=f.read(), time_modified=_boto3_now())
 
     def __getattr__(self, key):
         if key in ('e_tag', 'last_modified', 'size'):
@@ -303,7 +306,9 @@ class MockS3Object(object):
         if self.key not in mock_keys:
             raise _no_such_key_error(self.key, 'GetObject')
 
-        return mock_keys[self.key]
+        key_info = mock_keys[self.key]
+
+        return (key_info['body'], key_info['time_modified'])
 
 
 class MockStreamingBody(object):

--- a/tests/mock_boto3/s3.py
+++ b/tests/mock_boto3/s3.py
@@ -35,7 +35,8 @@ class MockS3Client(object):
                        and *time_modified* is a UTC
                        :py:class:`~datetime.datetime`. *location* is an
                        optional location constraint for the bucket
-                       (a region name).
+                       (a region name). *storage_class* is an optional
+                       non-Standard storage class (e.g. ``'GLACIER'``).
     """
     def __init__(self,
                  mock_s3_fs,
@@ -224,19 +225,24 @@ class MockS3Object(object):
         return {}
 
     def get(self):
-        key_data, mtime = self._get_key_data_and_mtime()
+        mock_keys = self._mock_bucket_keys('GetBucket')
+
+        if self.key not in mock_keys:
+            raise _no_such_key_error(self.key, 'GetObject')
+
+        mock_key = mock_keys[self.key]
 
         # fill in known attributes
         m = hashlib.md5()
-        m.update(key_data)
+        m.update(mock_key['body'])
 
         self.e_tag = '"%s"' % m.hexdigest()
-        self.last_modified = mtime
-        self.size = len(key_data)
-        self.storage_class = None
+        self.last_modified = mock_key['time_modified']
+        self.size = len(mock_key['body'])
+        self.storage_class = mock_key.get('storage_class')
 
         result = dict(
-            Body=MockStreamingBody(key_data),
+            Body=MockStreamingBody(mock_key['body']),
             ContentLength=self.size,
             ETag=self.e_tag,
             LastModified=self.last_modified,
@@ -298,17 +304,6 @@ class MockS3Object(object):
     def _check_bucket_exists(self, operation_name):
         if self.bucket_name not in self.meta.client.mock_s3_fs:
             raise _no_such_bucket_error(self.bucket_name, operation_name)
-
-    def _get_key_data_and_mtime(self):
-        """Return (key_data, time_modified)."""
-        mock_keys = self._mock_bucket_keys('GetBucket')
-
-        if self.key not in mock_keys:
-            raise _no_such_key_error(self.key, 'GetObject')
-
-        key_info = mock_keys[self.key]
-
-        return (key_info['body'], key_info['time_modified'])
 
 
 class MockStreamingBody(object):

--- a/tests/mock_boto3/s3.py
+++ b/tests/mock_boto3/s3.py
@@ -18,10 +18,12 @@ from datetime import datetime
 from datetime import timedelta
 
 from botocore.exceptions import ClientError
+from botocore.exceptions import ParamValidationError
 from boto3.exceptions import S3UploadFailedError
 
 from mrjob.aws import _DEFAULT_AWS_REGION
 from mrjob.aws import _boto3_now
+from mrjob.py2 import string_types
 
 from .util import MockClientMeta
 
@@ -214,6 +216,41 @@ class MockS3Object(object):
         self.key = key
 
         self.meta = MockClientMeta(client=client)
+
+    def copy_from(self, CopySource, StorageClass=None):
+        # very limited emulation of copy_from, to allow archiving
+        # in Glacier
+
+        # CopySource may be a string or a dict
+        if isinstance(CopySource, string_types):
+            bucket, key = CopySource.split('/', 1)
+            CopySource = dict(Bucket=bucket, Key=key)
+
+        # validate CopySource
+        if not (isinstance(CopySource, dict) and
+                set(CopySource) >= {'Bucket', 'Key'}):
+            # this isn't the exact error message
+            raise ParamValidationError(
+                report='CopySource must include Bucket and Key')
+
+        # look up the source key
+        mock_source_bucket = self.meta.client.mock_s3_fs.get(
+            CopySource['Bucket'])
+        if not mock_source_bucket:
+            raise _no_such_bucket_error(CopySource['Bucket'], 'CopyObject')
+
+        mock_source_key = mock_source_bucket['keys'].get(CopySource('Key'))
+        if not mock_source_key:
+            raise _no_such_key_error(CopySource['Key'], 'CopyObject')
+
+        new_mock_key = dict(mock_source_key)
+        new_mock_key['time_modified'] = _boto3_now()
+        new_mock_key['storage_class'] = StorageClass
+
+        mock_bucket_keys = self._mock_bucket_keys('CopyObject')
+        mock_bucket_keys[self.key] = new_mock_key
+
+        return {}
 
     def delete(self):
         mock_keys = self._mock_bucket_keys('DeleteObject')

--- a/tests/mock_boto3/s3.py
+++ b/tests/mock_boto3/s3.py
@@ -18,12 +18,10 @@ from datetime import datetime
 from datetime import timedelta
 
 from botocore.exceptions import ClientError
-from botocore.exceptions import ParamValidationError
 from boto3.exceptions import S3UploadFailedError
 
 from mrjob.aws import _DEFAULT_AWS_REGION
 from mrjob.aws import _boto3_now
-from mrjob.py2 import string_types
 
 from .util import MockClientMeta
 
@@ -223,41 +221,6 @@ class MockS3Object(object):
         self.key = key
 
         self.meta = MockClientMeta(client=client)
-
-    def copy_from(self, CopySource, StorageClass=None):
-        # very limited emulation of copy_from, to allow archiving
-        # in Glacier
-
-        # CopySource may be a string or a dict
-        if isinstance(CopySource, string_types):
-            bucket, key = CopySource.split('/', 1)
-            CopySource = dict(Bucket=bucket, Key=key)
-
-        # validate CopySource
-        if not (isinstance(CopySource, dict) and
-                set(CopySource) >= {'Bucket', 'Key'}):
-            # this isn't the exact error message
-            raise ParamValidationError(
-                report='CopySource must include Bucket and Key')
-
-        # look up the source key
-        mock_source_bucket = self.meta.client.mock_s3_fs.get(
-            CopySource['Bucket'])
-        if not mock_source_bucket:
-            raise _no_such_bucket_error(CopySource['Bucket'], 'CopyObject')
-
-        mock_source_key = mock_source_bucket['keys'].get(CopySource('Key'))
-        if not mock_source_key:
-            raise _no_such_key_error(CopySource['Key'], 'CopyObject')
-
-        new_mock_key = dict(mock_source_key)
-        new_mock_key['time_modified'] = _boto3_now()
-        new_mock_key['storage_class'] = StorageClass
-
-        mock_bucket_keys = self._mock_bucket_keys('CopyObject')
-        mock_bucket_keys[self.key] = new_mock_key
-
-        return {}
 
     def delete(self):
         mock_keys = self._mock_bucket_keys('DeleteObject')

--- a/tests/mock_boto3/s3.py
+++ b/tests/mock_boto3/s3.py
@@ -232,13 +232,19 @@ class MockS3Object(object):
         self.e_tag = '"%s"' % m.hexdigest()
         self.last_modified = mtime
         self.size = len(key_data)
+        self.storage_class = None
 
-        return dict(
+        result = dict(
             Body=MockStreamingBody(key_data),
             ContentLength=self.size,
             ETag=self.e_tag,
             LastModified=self.last_modified,
         )
+
+        if self.storage_class is not None:
+            result['StorageClass'] = self.storage_class
+
+        return result
 
     def put(self, Body):
         if not isinstance(Body, bytes):

--- a/tests/mock_boto3/s3.py
+++ b/tests/mock_boto3/s3.py
@@ -93,11 +93,13 @@ class MockS3Client(object):
         return dict(Buckets=buckets)
 
 
-def add_mock_s3_data(mock_s3_fs, data, age=None, location=None):
+def add_mock_s3_data(mock_s3_fs, data,
+                     age=None, location=None, storage_class=None):
     """Update *mock_s3_fs* with a map from bucket name to key name to data.
 
     :param age: a timedelta
     :param location string: the bucket's location constraint (a region name)
+    :param storage_class string: storage class for all data added
     """
     age = age or timedelta(0)
     time_modified = _boto3_now() - age
@@ -110,8 +112,13 @@ def add_mock_s3_data(mock_s3_fs, data, age=None, location=None):
         for key_name, key_data in key_name_to_bytes.items():
             if not isinstance(key_data, bytes):
                 raise TypeError('mock s3 data must be bytes')
-            bucket['keys'][key_name] = dict(
+
+            mock_key = dict(
                 body=key_data, time_modified=time_modified)
+            if storage_class:
+                mock_key['storage_class'] = storage_class
+
+            bucket['keys'][key_name] = mock_key
 
         if location is not None:
             bucket['location'] = location

--- a/tests/test_emr.py
+++ b/tests/test_emr.py
@@ -5662,6 +5662,25 @@ class CheckInputPathsTestCase(MockBoto3TestCase):
         with job.make_runner() as runner:
             self.assertRaises(StopIteration, runner.run)
 
+    def test_existing_s3_dir(self):
+        self.add_mock_s3_data({'walrus': {'data/one': b'one\n',
+                                          'data/two': b'two\n'}})
+
+        job = MRTwoStepJob(['-r', 'emr', 's3://walrus/data/'])
+
+        with job.make_runner() as runner:
+            self.assertRaises(StopIteration, runner.run)
+
+    def test_s3_path_in_glacier(self):
+        self.add_mock_s3_data({'walrus': {'data/one': b'one\n'}})
+        self.add_mock_s3_data({'walrus': {'data/two': b'two\n'}},
+                              storage_class='GLACIER')
+
+        job = MRTwoStepJob(['-r', 'emr', 's3://walrus/data/'])
+
+        with job.make_runner() as runner:
+            self.assertRaises(IOError, runner.run)
+
     def test_nonexistent_s3_path(self):
         job = MRTwoStepJob(['-r', 'emr', 's3://walrus/data/foo'])
 

